test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -779,27 +803,35 @@ void UcxConnection::accept(ucp_conn_request_h conn_req, UcxCallback *callback)
 
 void UcxConnection::disconnect(UcxCallback *callback)
 {
-    /* establish cb must be destroyed earlier since it accesses
-     * the connection */
-    assert(_establish_cb == NULL);
+    UCX_CONN_LOG << "disconnect, ep is " << _ep;
+
     assert(_disconnect_cb == NULL);
+    _disconnect_cb = callback;
 
-    UCX_CONN_LOG << "destroying, ep is " << _ep;
-    ep_close(UCP_EP_CLOSE_MODE_FORCE);
+    if (!UCS_STATUS_IS_ERR(_ucx_status)) {
+        _context.remove_connection(this);
 
-    _disconnect_cb = callback;
-    if (ucs_list_is_empty(&_all_requests)) {
-        _context.move_connection_to_disconnecting(this);
-    } else {
-        cancel_all();
+        if (!is_established()) {
+            assert(_ucx_status == UCS_INPROGRESS);
+            established(UCS_ERR_CANCELED);
+        }
     }
+
+    _context.move_connection_to_disconnecting(this);
+
+    cancel_all();
+    ep_close(UCP_EP_CLOSE_MODE_FORCE);
```

Comment:
disconnect_progress should call ep_close after ucs_list_is_empty is true , and only the first time (if _ep still non-NULL). remove assert in line 828.

Reply:
is it needed to purge requests? i guess will need to change it when porting to integration3?
pls add code comment why need to call ep_close here

Reply:
> is it needed to purge requests?

yes, it is needed to purge all outstanding requests.

> i guess will need to change it when porting to integration3?

yes, if it is not allowed to call ep_close() for EP with some outstanding requests, then we should wait for all requests prior closing EP.

> pls add code comment why need to call ep_close here

done